packages/preview2-shim/lib/io/worker-thread.js

```diff
@@ -0,0 +1,312 @@
+    case calls.OUTPUT_STREAM_CREATE | streamTypes.STDERR:
+      unfinishedStreams.set(++streamCnt, {
+        flushPromise: null,
+        stream: process.stderr,
+        blocksMainThread: true
+      });
+      return streamCnt;
+
+    // Filesystem
+    case calls.INPUT_STREAM_CREATE | streamTypes.FILE:
+      var { fd, offset } = payload;
+      var stream = fs.createReadStream(null, { fd, autoClose: false, start: Number(offset) });
+      // Node.js needs an initial empty read to start actually reading
+      if (stream.read())
+        throw new Error('Internal error unexpected data');
+      unfinishedStreams.set(++streamCnt, {
+        flushPromise: null,
+        stream,
+        blocksMainThread: false,
+      });
+      return streamCnt;
+    case calls.OUTPUT_STREAM_CREATE | streamTypes.FILE: throw new Error('todo: file write');
+
+    // Generic call implementations (streams + polls)
+    default: switch (call & calls.CALL_MASK) {
+      case calls.INPUT_STREAM_READ:
+        var { stream } = getStreamOrThrow(id);
+        return stream.read(Number(payload)) ?? new Uint8Array();
+      case calls.INPUT_STREAM_BLOCKING_READ:
+        return Promise.resolve(
+          unfinishedPolls.get(handle(calls.INPUT_STREAM_SUBSCRIBE | (call & calls.CALL_TYPE_MASK), id))
+        )
+        .then(() => handle(calls.INPUT_STREAM_READ | (call & calls.CALL_TYPE_MASK), id, payload));
+      case calls.INPUT_STREAM_SKIP:
+        return handle(calls.INPUT_STREAM_READ | (call & calls.CALL_TYPE_MASK), id, new Uint8Array(Number(payload)));
+      case calls.INPUT_STREAM_BLOCKING_SKIP:
+        return handle(calls.INPUT_STREAM_BLOCKING_READ | (call & calls.CALL_TYPE_MASK), id, new Uint8Array(Number(payload)));
+      case calls.INPUT_STREAM_SUBSCRIBE:
+        var stream = unfinishedStreams.get(id)?.stream;
+        // already closed or errored -> immediately return poll
+        // (poll 0 is immediately resolved)
+        if (!stream || stream.closed || stream.errored || stream.readable) return 0;
+        var resolve, reject;
+        // TODO: can we do this better with a single listener setup on stream
+        // creation to track the lifecycle promise?
+        unfinishedPolls.set(++pollCnt, new Promise((_resolve, _reject) => {
+          stream
+            .once('readable', resolve = _resolve)
+            .once('error', reject = _reject);
+        }).then(() => {
+          unfinishedPolls.delete(pollCnt);
+          stream
+            .off('error', reject)
+            .off('readable', resolve);
+        }, err => {
+          stream
+            .off('error', reject)
+            .off('readable', resolve);
+          unfinishedPolls.delete(pollCnt);
+          throw streamError(id, stream.stream, err);
+        }));
+        return pollCnt;
+      case calls.INPUT_STREAM_DROP:
+        unfinishedStreams.delete(id);
+        return;
+
+      case calls.OUTPUT_STREAM_CHECK_WRITE:
+        var { stream } = getStreamOrThrow(id);
```

Comment:
I've noticed `var` is being used a lot here, we could limit the scope of declared variables to the current lexical env using `const` or `let` to avoid any potential hoisting issues in the future - unless there is a valid reason I am not aware of.

```suggestion
        const { stream } = getStreamOrThrow(id);
```

Reply:
The reason I'm using vars in the switch statement is because I don't want to create a new block statement for every single case. We could add block statements though for sure.

Reply:
I've gone ahead and updating to using blocks.